pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+	/// pallet in a valid state from which another 'step' of migration work can be performed. In no
+	/// case should a step consume more than `available_weight`.
+	///
+	/// This should return a perbill indicating the aggregate progress of the migration. If
+	/// `Perbill::one()` is returned, the migration is considered complete and no further calls to
+	/// `step()` will be made. Any value less than `Perbill::one()` will result in another future
+	/// call to `step()`. Indeed, values < 1 are arbitrary, but the intent is to indicate progress
+	/// (so they should at least be monotonically increasing).
+	fn step(&self, previous_progress: Perbill, available_weight: Weight) -> (Perbill, Weight);
+}
+
+/// Our list of migrations. Any ordering considerations can be specified here (?).
+const MIGRATIONS: [&dyn Migration; 3] = [
+	&migrations::MM_001_AuthorMappingAddDeposit {},
+	&migrations::MM_002_StakingFixTotalBalance {},
+	&migrations::MM_003_StakingTransitionBoundedSet {},
+];
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	#[allow(unused_imports)] // TODO: why does it detect this as unused?
+	use sp_std::prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	#[pallet::event]
+	pub enum Event<T: Config> {
+		// e.g. runtime upgrade started, completed, etc.
+		RuntimeUpgradeStarted,
+		RuntimeUpgradeCompleted,
+		MigrationStarted(String),
+		MigrationProgress(String, Perbill),
+		MigrationCompleted(String),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+
+		/// on_runtime_upgrade is expected to be called exactly once after a runtime upgrade.
+		/// We use this as a chance to flag that we are now in upgrade-mode and begin our
+		/// migrations.
+		/// 
+		/// In the event that a migration is expected to take more than one block, ongoing migration
+		/// work could continue from block-to-block in this pallet's on_initialize function.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			// start by flagging that we are not fully upgraded
+			<FullyUpgraded<T>>::put(false);
+
+			let mut weight: Weight = 0u64.into();
+
+			weight += process_runtime_upgrades::<T>();
+
+			weight.into()
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	type MigrationState<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		String,
+		Perbill,
+		OptionQuery, // TODO: what is this...?
```

Comment:
`OptionQuery` means its a map from `Key => Option<Value>` i.e. `String => Option<Perbill>` for this map

`ValueQuery` means its a map from `Key => Value` such that the default Value for all unset Keys is `Value::default()`